java.lang.NullPointerException: Cannot read field "declaringScope" because "outerLocalVariable" is null after upgrading from 2.11 to 2.12.1

GWT version:2.12.1 

After upgrading gwt in our application from 2.11 to 2.12.1 we get the below error during the gwt compilation process.
 java.lang.NullPointerException: Cannot read field "declaringScope" because "outerLocalVariable" is null

The full stack trace:
[INFO] [ERROR] Unexpected internal compiler error
[INFO] java.lang.NullPointerException: Cannot read field "declaringScope" because "outerLocalVariable" is null
[INFO]  at org.eclipse.jdt.internal.compiler.lookup.BlockScope.getEmulationPath(BlockScope.java:831)
[INFO]  at com.google.gwt.dev.jjs.impl.GwtAstBuilder$AstVisitor.pushLambdaExpressionLocalsIntoMethodScope(GwtAstBuilder.java:1264)
[INFO]  at com.google.gwt.dev.jjs.impl.GwtAstBuilder$AstVisitor.visit(GwtAstBuilder.java:1253)
[INFO]  at org.eclipse.jdt.internal.compiler.ast.LambdaExpression.traverse(LambdaExpression.java:787)
[INFO]  at org.eclipse.jdt.internal.compiler.ast.ExplicitConstructorCall.traverse(ExplicitConstructorCall.java:524)
[INFO]  at org.eclipse.jdt.internal.compiler.ast.ConstructorDeclaration.traverse(ConstructorDeclaration.java:704)
[INFO]  at org.eclipse.jdt.internal.compiler.ast.TypeDeclaration.traverse(TypeDeclaration.java:1831)
[INFO]  at org.eclipse.jdt.internal.compiler.ast.TypeDeclaration.traverse(TypeDeclaration.java:1683)
[INFO]  at com.google.gwt.dev.jjs.impl.GwtAstBuilder.processImpl(GwtAstBuilder.java:4122)
[INFO]  at com.google.gwt.dev.jjs.impl.GwtAstBuilder.process(GwtAstBuilder.java:4160)
[INFO]  at com.google.gwt.dev.javac.CompilationStateBuilder$CompileMoreLater$UnitProcessorImpl.process(CompilationStateBuilder.java:128)
[INFO]  at com.google.gwt.dev.javac.JdtCompiler$CompilerImpl.process(JdtCompiler.java:322)
[INFO]  at org.eclipse.jdt.internal.compiler.Compiler.processCompiledUnits(Compiler.java:575)
[INFO]  at org.eclipse.jdt.internal.compiler.Compiler.compile(Compiler.java:475)
[INFO]  at org.eclipse.jdt.internal.compiler.Compiler.compile(Compiler.java:426)
[INFO]  at com.google.gwt.dev.javac.JdtCompiler.doCompile(JdtCompiler.java:1021)
[INFO]  at com.google.gwt.dev.javac.CompilationStateBuilder$CompileMoreLater.compile(CompilationStateBuilder.java:322)
[INFO]  at com.google.gwt.dev.javac.CompilationStateBuilder.doBuildFrom(CompilationStateBuilder.java:532)
[INFO]  at com.google.gwt.dev.javac.CompilationStateBuilder.buildFrom(CompilationStateBuilder.java:464)
[INFO]  at com.google.gwt.dev.cfg.ModuleDef.getCompilationState(ModuleDef.java:426)
[INFO]  at com.google.gwt.dev.Precompile.precompile(Precompile.java:210)
[INFO]  at com.google.gwt.dev.Precompile.precompile(Precompile.java:190)
[INFO]  at com.google.gwt.dev.Precompile.precompile(Precompile.java:131)
[INFO]  at com.google.gwt.dev.Compiler.compile(Compiler.java:192)
[INFO]  at com.google.gwt.dev.Compiler.compile(Compiler.java:143)
[INFO]  at com.google.gwt.dev.Compiler.compile(Compiler.java:132)
[INFO]  at com.google.gwt.dev.Compiler$1.run(Compiler.java:110)
[INFO]  at com.google.gwt.dev.CompileTaskRunner.doRun(CompileTaskRunner.java:55)
[INFO]  at com.google.gwt.dev.CompileTaskRunner.runWithAppropriateLogger(CompileTaskRunner.java:50)
[INFO]  at com.google.gwt.dev.Compiler.main(Compiler.java:113)

Hi Colin,

thank you for this version. I've testen and the compilation failure has more details.

[INFO] [ERROR] An internal compiler exception occurred
[INFO] com.google.gwt.dev.jjs.InternalCompilerException: Error constructing Java AST
[INFO]  at com.google.gwt.dev.jjs.impl.GwtAstBuilder.translateException(GwtAstBuilder.java:4190)
[INFO]  at com.google.gwt.dev.jjs.impl.GwtAstBuilder$AstVisitor.visit(GwtAstBuilder.java:1257)
[INFO]  at org.eclipse.jdt.internal.compiler.ast.LambdaExpression.traverse(LambdaExpression.java:787)
[INFO]  at org.eclipse.jdt.internal.compiler.ast.ExplicitConstructorCall.traverse(ExplicitConstructorCall.java:524)
[INFO]  at org.eclipse.jdt.internal.compiler.ast.ConstructorDeclaration.traverse(ConstructorDeclaration.java:704)
[INFO]  at org.eclipse.jdt.internal.compiler.ast.TypeDeclaration.traverse(TypeDeclaration.java:1831)
[INFO]  at org.eclipse.jdt.internal.compiler.ast.TypeDeclaration.traverse(TypeDeclaration.java:1683)
[INFO]  at com.google.gwt.dev.jjs.impl.GwtAstBuilder.processImpl(GwtAstBuilder.java:4126)
[INFO]  at com.google.gwt.dev.jjs.impl.GwtAstBuilder.process(GwtAstBuilder.java:4164)
[INFO]  at com.google.gwt.dev.javac.CompilationStateBuilder$CompileMoreLater$UnitProcessorImpl.process(CompilationStateBuilder.java:128)
[INFO]  at com.google.gwt.dev.javac.JdtCompiler$CompilerImpl.process(JdtCompiler.java:322)
[INFO]  at org.eclipse.jdt.internal.compiler.Compiler.processCompiledUnits(Compiler.java:575)
[INFO]  at org.eclipse.jdt.internal.compiler.Compiler.compile(Compiler.java:475)
[INFO]  at org.eclipse.jdt.internal.compiler.Compiler.compile(Compiler.java:426)
[INFO]  at com.google.gwt.dev.javac.JdtCompiler.doCompile(JdtCompiler.java:1021)
[INFO]  at com.google.gwt.dev.javac.CompilationStateBuilder$CompileMoreLater.compile(CompilationStateBuilder.java:322)
[INFO]  at com.google.gwt.dev.javac.CompilationStateBuilder.doBuildFrom(CompilationStateBuilder.java:532)
[INFO]  at com.google.gwt.dev.javac.CompilationStateBuilder.buildFrom(CompilationStateBuilder.java:464)
[INFO]  at com.google.gwt.dev.cfg.ModuleDef.getCompilationState(ModuleDef.java:426)
[INFO]  at com.google.gwt.dev.Precompile.precompile(Precompile.java:210)
[INFO]  at com.google.gwt.dev.Precompile.precompile(Precompile.java:190)
[INFO]  at com.google.gwt.dev.Precompile.precompile(Precompile.java:131)
[INFO]  at com.google.gwt.dev.Compiler.compile(Compiler.java:192)
[INFO]  at com.google.gwt.dev.Compiler.compile(Compiler.java:143)
[INFO]  at com.google.gwt.dev.Compiler.compile(Compiler.java:132)
[INFO]  at com.google.gwt.dev.Compiler$1.run(Compiler.java:110)
[INFO]  at com.google.gwt.dev.CompileTaskRunner.doRun(CompileTaskRunner.java:55)
[INFO]  at com.google.gwt.dev.CompileTaskRunner.runWithAppropriateLogger(CompileTaskRunner.java:50)
[INFO]  at com.google.gwt.dev.Compiler.main(Compiler.java:113)
[INFO] Caused by: java.lang.NullPointerException: Cannot read field "declaringScope" because "outerLocalVariable" is null
[INFO]  at org.eclipse.jdt.internal.compiler.lookup.BlockScope.getEmulationPath(BlockScope.java:831)
[INFO]  at com.google.gwt.dev.jjs.impl.GwtAstBuilder$AstVisitor.pushLambdaExpressionLocalsIntoMethodScope(GwtAstBuilder.java:1268)
[INFO]  at com.google.gwt.dev.jjs.impl.GwtAstBuilder$AstVisitor.visit(GwtAstBuilder.java:1254)
[INFO]  ... 27 more
[INFO]    [ERROR] at PdfVersiesView.java(273): (<no type> afdruk) -> {
[INFO]  //...
[INFO] }
[INFO]       org.eclipse.jdt.internal.compiler.ast.LambdaExpression
[INFO] ------------------------------------------------------------------------

It looks like the compiler is complaining about the "<no type> " of afdruk variable.

The actual code is 
`private class DownloadCell extends ActionImageResourceCell<AfdrukMetaDataResource> {
		public DownloadCell() {
			super(afdruk -> {
				//...
			});
		}`

After changing it to the below code (by not using lambda) resolves the compilation problem .
`private class DownloadCell extends ActionImageResourceCell<AfdrukMetaDataResource> {
		public DownloadCell() {
			super(new Delegate<AfdrukMetaDataResource>() {
				@Override
				public void execute(AfdrukMetaDataResource afdruk) {
					//...
				}
			});
		}`

The parent class ActionImageResourceCell looks like this

`public abstract class ActionImageResourceCell<C> extends ActionCell<C> {
	public ActionImageResourceCell(ActionCell.Delegate<C> delegate) {
		super("(ignored)", delegate);
	}
`

Regards, Koen

A week's worth of notes from puttering with this:

I've ruled out (at least for the moment) applying the same change in other places, as it introduces other problems where we expect the same synthetic local binding instance that was previously visited - it isn't necessarily required to make this change, but I'm worried about other ways that this bug can happen.

I am however getting compiler errors with the same minimal test case in GWT 2.11:

```
     [java]    [ERROR] Errors in 'file:/home/colin/workspace/gwt/samples/hello/src/com/google/gwt/sample/hello/client/TestComposite.java'
     [java]       [ERROR] Line 16: Cannot refer to 'this' nor 'super' while explicitly invoking a constructor
```
This sort of makes sense to me, if this error was previously (incorrectly) being emitted by JDT, then updating the JDT changed so that we got this current bug instead.

@koendecock can you confirm that the minimal example you gave succeeds in 2.11 for you?

---

Focusing only on the code that breaks GWT 2.12 with this stack trace, I further reduced the example - a lambda in a non-static inner class in an outer class with a field. The lambda must be declared in a constructor and close over a _field_ in the outermost class, and passed to super()/this():
```java
public class OuterClass {
  boolean outerField = false;

  public static class SuperInnerType {
    public SuperInnerType(Runnable delegate) {
    }
  }

  class NonStaticInnerType extends SuperInnerType {
    NonStaticInnerType() {
      // Constructor of a non-static inner class passes a lambda to super() or this(), and
      // that lambda captures a field of the outer class.
      super(() -> {
        outerField = true;
      });
    }
  }
}
```
or
```java
public class OuterClass {
  boolean outerField = false;

  class NonStaticInnerType {
    NonStaticInnerType() {
      // Constructor of a non-static inner class passes a lambda to super() or this(), and
      // that lambda captures a field of the outer class.
      this(() -> {
        outerField = true;
      });
    }
    NonStaticInnerType(Runnable delegate) {
    }
  }
}
```
Note that both of these appear to **compile cleanly** in GWT 2.11, and fail in GWT 2.12 as reported.

However, both of these following two tests (written for `Java8AstTest`) fail in GWT 2.12 (with the NPE), and fail in GWT 2.11 (with the "Cannot refer to 'this' or 'super'" error):
```
  public void testCompileLambdaOuterFieldCaptureInConstructor() throws UnableToCompleteException {
    addSnippetClassDecl("int y = 22;");
    addSnippetClassDecl("class Foo {" +
        "Foo(Runnable r) {}" +
        "Foo() {" +
            "this(() -> {y = 42;});" +
        "}" +
    "}");

    JProgram program = compileSnippet("void", "new Foo();", false);
    assertNotNull(getMethod(program, "lambda$0"));
  }

  public void testCompileLambdaOuterFieldCaptureInConstructorSuper() throws UnableToCompleteException {
    addSnippetClassDecl("int y = 22;");
    addSnippetClassDecl("class Foo {" +
        "Foo(Runnable r) {}" +
    "}" +
    "class Bar extends Foo {" +
        "Bar() {" +
            "super(() -> {y = 42;});" +
        "}" +
    "}");

    JProgram program = compileSnippet("void", "new Bar();", false);
    assertNotNull(getMethod(program, "lambda$0"));
  }
```

Why does "OuterClass` succeed in both forms, but your sample and my unit test fail in GWT 2.11? I'm sure I'm just missing something totally unrelated here...

--- a/dev/core/src/com/google/gwt/dev/jjs/impl/GwtAstBuilder.java
+++ b/dev/core/src/com/google/gwt/dev/jjs/impl/GwtAstBuilder.java
@@ -3612,7 +3612,7 @@ public class GwtAstBuilder {
         // Synthetic locals for local classes
         if (targetBinding.syntheticOuterLocalVariables() != null) {
           for (SyntheticArgumentBinding arg : targetBinding.syntheticOuterLocalVariables()) {
-            LocalVariableBinding targetVariable = arg.actualOuterLocalVariable;
+            LocalVariableBinding targetVariable = arg.actualOuterLocalVariable;// this may be wrong too
             VariableBinding[] path = scope.getEmulationPath(targetVariable);
             assert path.length == 1;
             if (curMethod.scope.isInsideInitializer()

--- a/dev/core/src/com/google/gwt/dev/jjs/impl/GwtAstBuilder.java
+++ b/dev/core/src/com/google/gwt/dev/jjs/impl/GwtAstBuilder.java
@@ -1237,31 +1237,27 @@ public class GwtAstBuilder {
 
     @Override
     public boolean visit(LambdaExpression x, BlockScope blockScope) {
-      try {
-        // Fetch the variables 'captured' by this lambda
-        SyntheticArgumentBinding[] synthArgs = x.outerLocalVariables;
-        // Get the parameter names, captured locals + lambda arguments
-        String paramNames[] = computeCombinedParamNames(x, synthArgs);
-        SourceInfo info = makeSourceInfo(x);
-        // JDT synthesizes a method lambda$n(capture1, capture2, ..., lambda_arg1, lambda_arg2, ...)
-        // Here we create a JMethod from this
-        JMethod lambdaMethod = createMethodFromBinding(info, x.binding, paramNames);
-        // Because the lambda implementations is synthesized as a static method in the
-        // enclosing class, it needs to be adjusted if that class happens to be a JsType.
-        lambdaMethod.setJsMemberInfo(JsMemberType.NONE, null, null, false);
-        if (curClass.type.isJsNative()) {
-          lambdaMethod.setJsOverlay();
-        }
-        JMethodBody methodBody = new JMethodBody(info);
-        lambdaMethod.setBody(methodBody);
-        // We need to push this method  on the stack as it introduces a scope, and
-        // expressions in the body need to lookup variable refs like parameters from it
-        pushMethodInfo(new MethodInfo(lambdaMethod, methodBody, x.scope));
-        pushLambdaExpressionLocalsIntoMethodScope(x, synthArgs, lambdaMethod);
-        // now the body of the lambda is processed
-      } catch (Throwable e) {
-        throw translateException(x, e);
+      // Fetch the variables 'captured' by this lambda
+      SyntheticArgumentBinding[] synthArgs = x.outerLocalVariables;
+      // Get the parameter names, captured locals + lambda arguments
+      String paramNames[] = computeCombinedParamNames(x, synthArgs);
+      SourceInfo info = makeSourceInfo(x);
+      // JDT synthesizes a method lambda$n(capture1, capture2, ..., lambda_arg1, lambda_arg2, ...)
+      // Here we create a JMethod from this
+      JMethod lambdaMethod = createMethodFromBinding(info, x.binding, paramNames);
+      // Because the lambda implementations is synthesized as a static method in the
+      // enclosing class, it needs to be adjusted if that class happens to be a JsType.
+      lambdaMethod.setJsMemberInfo(HasJsInfo.JsMemberType.NONE, null, null, false);
+      if (curClass.type.isJsNative()) {
+        lambdaMethod.setJsOverlay();
       }
+      JMethodBody methodBody = new JMethodBody(info);
+      lambdaMethod.setBody(methodBody);
+      // We need to push this method  on the stack as it introduces a scope, and
+      // expressions in the body need to lookup variable refs like parameters from it
+      pushMethodInfo(new MethodInfo(lambdaMethod, methodBody, x.scope));
+      pushLambdaExpressionLocalsIntoMethodScope(x, synthArgs, lambdaMethod);
+      // now the body of the lambda is processed
       return true;
     }
 

--- a/dev/core/src/com/google/gwt/dev/jjs/impl/GwtAstBuilder.java
+++ b/dev/core/src/com/google/gwt/dev/jjs/impl/GwtAstBuilder.java
@@ -1267,7 +1267,12 @@ public class GwtAstBuilder {
       if (syntheticArguments != null) {
         MethodScope scope = x.getScope();
         for (SyntheticArgumentBinding sa : syntheticArguments) {
-          VariableBinding[] path = scope.getEmulationPath(sa);
+          VariableBinding[] path;
+          if (sa.actualOuterLocalVariable == null) {
+            path = scope.getEmulationPath(sa);
+          } else {
+            path = scope.getEmulationPath(sa.actualOuterLocalVariable);
+          }
           assert path.length == 1 && path[0] instanceof LocalVariableBinding;
           JParameter param = it.next();
           curMethod.locals.put((LocalVariableBinding) path[0], param);

--- a/dev/core/src/com/google/gwt/dev/jjs/impl/GwtAstBuilder.java
+++ b/dev/core/src/com/google/gwt/dev/jjs/impl/GwtAstBuilder.java
@@ -1237,27 +1237,31 @@ public class GwtAstBuilder {
 
     @Override
     public boolean visit(LambdaExpression x, BlockScope blockScope) {
-      // Fetch the variables 'captured' by this lambda
-      SyntheticArgumentBinding[] synthArgs = x.outerLocalVariables;
-      // Get the parameter names, captured locals + lambda arguments
-      String paramNames[] = computeCombinedParamNames(x, synthArgs);
-      SourceInfo info = makeSourceInfo(x);
-      // JDT synthesizes a method lambda$n(capture1, capture2, ..., lambda_arg1, lambda_arg2, ...)
-      // Here we create a JMethod from this
-      JMethod lambdaMethod = createMethodFromBinding(info, x.binding, paramNames);
-      // Because the lambda implementations is synthesized as a static method in the
-      // enclosing class, it needs to be adjusted if that class happens to be a JsType.
-      lambdaMethod.setJsMemberInfo(HasJsInfo.JsMemberType.NONE, null, null, false);
-      if (curClass.type.isJsNative()) {
-        lambdaMethod.setJsOverlay();
+      try {
+        // Fetch the variables 'captured' by this lambda
+        SyntheticArgumentBinding[] synthArgs = x.outerLocalVariables;
+        // Get the parameter names, captured locals + lambda arguments
+        String paramNames[] = computeCombinedParamNames(x, synthArgs);
+        SourceInfo info = makeSourceInfo(x);
+        // JDT synthesizes a method lambda$n(capture1, capture2, ..., lambda_arg1, lambda_arg2, ...)
+        // Here we create a JMethod from this
+        JMethod lambdaMethod = createMethodFromBinding(info, x.binding, paramNames);
+        // Because the lambda implementations is synthesized as a static method in the
+        // enclosing class, it needs to be adjusted if that class happens to be a JsType.
+        lambdaMethod.setJsMemberInfo(JsMemberType.NONE, null, null, false);
+        if (curClass.type.isJsNative()) {
+          lambdaMethod.setJsOverlay();
+        }
+        JMethodBody methodBody = new JMethodBody(info);
+        lambdaMethod.setBody(methodBody);
+        // We need to push this method  on the stack as it introduces a scope, and
+        // expressions in the body need to lookup variable refs like parameters from it
+        pushMethodInfo(new MethodInfo(lambdaMethod, methodBody, x.scope));
+        pushLambdaExpressionLocalsIntoMethodScope(x, synthArgs, lambdaMethod);
+        // now the body of the lambda is processed
+      } catch (Throwable e) {
+        throw translateException(x, e);
       }
-      JMethodBody methodBody = new JMethodBody(info);
-      lambdaMethod.setBody(methodBody);
-      // We need to push this method  on the stack as it introduces a scope, and
-      // expressions in the body need to lookup variable refs like parameters from it
-      pushMethodInfo(new MethodInfo(lambdaMethod, methodBody, x.scope));
-      pushLambdaExpressionLocalsIntoMethodScope(x, synthArgs, lambdaMethod);
-      // now the body of the lambda is processed
       return true;
     }
 

--- a/dev/core/src/com/google/gwt/dev/jjs/impl/GwtAstBuilder.java
+++ b/dev/core/src/com/google/gwt/dev/jjs/impl/GwtAstBuilder.java
@@ -3612,7 +3612,7 @@ public class GwtAstBuilder {
         // Synthetic locals for local classes
         if (targetBinding.syntheticOuterLocalVariables() != null) {
           for (SyntheticArgumentBinding arg : targetBinding.syntheticOuterLocalVariables()) {
-            LocalVariableBinding targetVariable = arg.actualOuterLocalVariable;// this may be wrong too
+            LocalVariableBinding targetVariable = arg.actualOuterLocalVariable;
             VariableBinding[] path = scope.getEmulationPath(targetVariable);
             assert path.length == 1;
             if (curMethod.scope.isInsideInitializer()

--- a/dev/core/src/com/google/gwt/dev/jjs/impl/GwtAstBuilder.java
+++ b/dev/core/src/com/google/gwt/dev/jjs/impl/GwtAstBuilder.java
@@ -1271,7 +1271,7 @@ public class GwtAstBuilder {
       if (syntheticArguments != null) {
         MethodScope scope = x.getScope();
         for (SyntheticArgumentBinding sa : syntheticArguments) {
-          VariableBinding[] path = scope.getEmulationPath(sa.actualOuterLocalVariable);
+          VariableBinding[] path = scope.getEmulationPath(sa);
           assert path.length == 1 && path[0] instanceof LocalVariableBinding;
           JParameter param = it.next();
           curMethod.locals.put((LocalVariableBinding) path[0], param);

--- a/dev/core/test/com/google/gwt/dev/jjs/impl/Java8AstTest.java
+++ b/dev/core/test/com/google/gwt/dev/jjs/impl/Java8AstTest.java
@@ -257,6 +257,88 @@ public class Java8AstTest extends FullCompileTestBase {
         formatSource(samMethod.toSource()));
   }
 
+  public void testCompileLambdaOuterFieldCaptureInConstructor() throws Exception {
+    addSnippetClassDecl("int y = 22;");
+    addSnippetClassDecl("class Foo {" +
+        "Foo(Runnable r) {}" +
+        "Foo() {" +
+            "this(() -> {y = 42;});" +
+        "}" +
+    "}");
+
+    JProgram program = compileSnippet("void", "new Foo();", false);
+    JDeclaredType entrypointType = program.getFromTypeMap("test.EntryPoint");
+
+    JMethod lambda = findMethod(program.getFromTypeMap("test.EntryPoint$Foo"), "lambda$0");
+    assertNotNull(lambda);
+    assertEquals("{{this$0.y=42;}}", formatSource(lambda.getBody().toSource()));
+
+    // Lambda closes over inner class's this$0 reference - lambda type constructor takes that
+    // parameter and assigns to a field.
+    JClassType lambdaInnerClass = (JClassType) getType(program, "test.EntryPoint$Foo$lambda$0$Type");
+    assertNotNull(lambdaInnerClass);
+
+    JMethod ctor = findMethod(lambdaInnerClass, "EntryPoint$Foo$lambda$0$Type");
+    assertTrue(ctor instanceof JConstructor);
+    System.out.println(ctor.getOriginalParamTypes());
+    assertEquals(1, ctor.getParams().size());
+    assertEquals(entrypointType, ctor.getOriginalParamTypes().get(0));
+
+    assertEquals(1, lambdaInnerClass.getFields().size());
+    assertEquals(entrypointType, lambdaInnerClass.getFields().get(0).getType());
+
+    // interface impl passes this.this$0 to the actual lambda method as a parameter
+    assertEquals("{this.this$0=this$0;}", formatSource(ctor.getBody().toSource()));
+
+    // abstract method implementation passes field to lambda method
+    JMethod samMethod = findMethod(lambdaInnerClass, "run");
+    assertNotNull(samMethod);
+    assertEquals("{EntryPoint$Foo.lambda$0(this.this$0);}", formatSource(samMethod.getBody().toSource()));
+  }
+
+  public void testCompileLambdaOuterFieldCaptureInConstructorSuper() throws Exception {
+    addSnippetClassDecl("int y = 22;");
+    addSnippetClassDecl("class Foo {" +
+        "Foo(Runnable r) {}" +
+    "}" +
+    "class Bar extends Foo {" +
+        "Bar() {" +
+            "super(() -> {y = 42;});" +
+        "}" +
+    "}");
+
+    JProgram program = compileSnippet("void", "new Bar();", false);
+    JDeclaredType entrypointType = program.getFromTypeMap("test.EntryPoint");
+
+    JMethod lambda = findMethod(program.getFromTypeMap("test.EntryPoint$Bar"), "lambda$0");
+    assertNotNull(lambda);
+    assertEquals(1, lambda.getParams().size());
+    assertEquals(entrypointType, lambda.getOriginalParamTypes().get(0));
+    assertEquals("{{this$0.y=42;}}", formatSource(lambda.getBody().toSource()));
+
+    // Lambda closes over inner class's this$0 reference - lambda type constructor takes that
+    // parameter and assigns to a field.
+    JClassType lambdaInnerClass = (JClassType) getType(program, "test.EntryPoint$Bar$lambda$0$Type");
+    assertNotNull(lambdaInnerClass);
+
+    JMethod ctor = findMethod(lambdaInnerClass, "EntryPoint$Bar$lambda$0$Type");
+    assertTrue(ctor instanceof JConstructor);
+    System.out.println(ctor.getOriginalParamTypes());
+    assertEquals(1, ctor.getParams().size());
+    assertEquals(entrypointType, ctor.getOriginalParamTypes().get(0));
+
+    assertEquals(1, lambdaInnerClass.getFields().size());
+    assertEquals(entrypointType, lambdaInnerClass.getFields().get(0).getType());
+
+    // interface impl passes this.this$0 to the actual lambda method as a parameter
+    assertEquals("{this.this$0=this$0;}", formatSource(ctor.getBody().toSource()));
+
+    // abstract method implementation passes field to lambda method
+    JMethod samMethod = findMethod(lambdaInnerClass, "run");
+    assertNotNull(samMethod);
+    assertEquals("{EntryPoint$Bar.lambda$0(this.this$0);}", formatSource(samMethod.getBody().toSource()));
+  }
+
   // make sure nested scoping of identically named variables works
   public void testCompileLambdaCaptureOuterInnerField() throws Exception {
     addSnippetClassDecl("private int y = 22;");

--- a/user/test/com/google/gwt/dev/jjs/test/Java8Test.java
+++ b/user/test/com/google/gwt/dev/jjs/test/Java8Test.java
@@ -252,6 +252,39 @@ public class Java8Test extends GWTTestCase {
     assertEquals(82, new AcceptsLambda<Integer>().accept(l).intValue());
   }
 
+  class CtorAcceptsLambda {
+    CtorAcceptsLambda() {
+      this(() -> local = -1);
+    }
+    CtorAcceptsLambda(Runnable lambda) {
+      lambda.run();
+    }
+  }
+
+  public void testCompileLambdaOuterFieldCaptureInConstructor() {
+    assertEquals(42, local);
+    new CtorAcceptsLambda();
+    assertEquals(-1, local);
+  }
+
+  abstract class AbstractCtorAcceptsLambda {
+    AbstractCtorAcceptsLambda(Runnable lambda) {
+      lambda.run();
+    }
+  }
+
+  class CtorAcceptsLambdaSubtype extends AbstractCtorAcceptsLambda {
+    CtorAcceptsLambdaSubtype() {
+      super(() -> local = -1);
+    }
+  }
+
+  public void testCompileLambdaOuterFieldCaptureInConstructorSuper() throws Exception {
+    assertEquals(42, local);
+    new CtorAcceptsLambdaSubtype();
+    assertEquals(-1, local);
+  }
+
   public void testCompileLambdaCaptureOuterInnerField() throws Exception {
     new Inner().run();
   }

